support-lib/cpp/SharedFuture.hpp

```diff
@@ -0,0 +1,138 @@
+/**
+ * Copyright 2021 Snap, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+ 
+ #pragma once
+
+#include "Future.hpp"
+
+#include <memory>
+#include <optional>
+#include <type_traits>
+#include <variant>
+#include <vector>
+
+namespace djinni {
+
+// SharedFuture is a wrapper around djinni::Future to allow multiple consumers (i.e. like std::shared_future)
+// The API is designed to be similar to djinni::Future.
+template<typename T>
+class SharedFuture {
+public:
+    // Create SharedFuture from Future. Runtime error if the future is already consumed.
+    explicit SharedFuture(Future<T>&& future);
+
+    // Transform into Future<T>.
+    Future<T> toFuture() const {
+        if (await_ready()) {
+            co_return await_resume(); // return stored value directly
+        }
+        co_return co_await SharedFuture(*this); // retain copy during coroutine suspension
+    }
+
+    void wait() const {
+        return [this]() -> Future<void> { co_await *this; }().wait();
+    }
+
+    decltype(auto) get() const {
+        wait();
+        return await_resume();
+    }
+
+    // Transform the result of this future into a new future. The behavior is same as Future::then except that
+    // it doesn't consume the future, and can be called multiple times.
+    template<typename Func>
+    SharedFuture<std::remove_cv_t<std::remove_reference_t<std::invoke_result_t<Func, T>>>> then(Func transform) const {
```

Comment:
yes this is a difference from Future, where you transform a ready future instead of the result.   This was intentional so that the transform function gets to handle exceptions.

Reply:
Good point regarding exception handling. Changed to handle similarly to Future.

> Question: is it valid to have this return a Future? Then it costs less resources by default and the user can decide whether to make it into a SharedFuture or not.

Since this class really is just for convenience, I would opt to keep this chaining to SharedFutures. I think it makes more sense semantically and is cleaner to produce more SharedFutures.  Otherwise, this could have just been a simple helper method `cloneFuture(Future<T>& f) -> Future<T>` instead of a full blown class.

Finally, regarding `remove_cvref_t`, that's what I used to have, but that is only available in C++20, and we are still configured as c++ 17 for now.

Reply:
> Since this class really is just for convenience, I would opt to keep this chaining to SharedFutures. I think it makes more sense semantically and is cleaner to produce more SharedFutures. Otherwise, this could have just been a simple helper method `cloneFuture(Future<T>& f) -> Future<T>` instead of a full blown class.

I think the advantage of returning a regular `Future` is that it a) uses less resources and b) can deal with move-only returnvalues. For example: a transform functor returning a `unique_ptr` will be extremely awkward if `then` returns a `SharedFuture`.

I expect that returning a `Future` will rarely, if ever, be inconvenient. Say I want to give a `Future` to multiple consumers. So I make it into a `SharedFuture` and hand it out. One of those consumers calls `then` on the `SharedFuture`. Will that consumer have multiple sub-consumers again? Maybe, but I'd expect the answer to be no most of the time. And when it does, it can always convert to a `SharedFuture` on the spot. You can always make a `Future` into a `SharedFuture`, but never the other way around.

That being said, I'm guessing you implemented this class based on a real use-case. I'd be very interested in an example or explanation of when it's inconvenient to return `Future` instead of `SharedFuture`. Maybe my expectations are not reflective of reality 🤷 

> Finally, regarding remove_cvref_t, that's what I used to have, but that is only available in C++20, and we are still configured as c++ 17 for now.

Good point, I wasn't aware that it's C++20 👍

Reply:
> I think the advantage of returning a regular Future is that it a) uses less resources and b) can deal with move-only returnvalues. For example: a transform functor returning a unique_ptr will be extremely awkward if then returns a SharedFuture.

I agree about resources.
But it turns out djnni::Future actually doesn't work with move-only types, anyway. At least that was the case last time I tried; is it different now?

> You can always make a Future into a SharedFuture, but never the other way around.

Well you can just call `toFuture()` on the result. it's just a tradeoff of resource vs. convenience.

> I'd be very interested in an example or explanation of when it's inconvenient to return Future instead of SharedFuture.

Sure. I added this class because it proves to be challenging to implement a djinni interface that is based on Futures. Specifically, you have a bunch of intermediate results that are Futures, and you need to apply some processing on them when the caller invokes methods returning other Futures. I practice, I always have a bunch of Futures that internally depend on other Futures, in the form of a DAG. This means I needed additional resolved result storage solution for each Future. SharedFuture solved this issue, and I needed all these intermediate results to be SharedFutures, as well. I only ever produce regular Futures again when I finally export them out of the djinni boundary to another language.

Reply:
> I agree about resources.
But it turns out djnni::Future actually doesn't work with move-only types, anyway. At least that was the case last time I tried; is it different now?

Since https://github.com/Snapchat/djinni/pull/180 it should. If not, then that's a bug that should be fixed. From a design standpoint, `Future` is able to deal with move-only types. `SharedFuture` is unable to deal well with move-only types by design, because it can't move them out of the shared state.

> Well you can just call toFuture() on the result. it's just a tradeoff of resource vs. convenience.

True, but that doesn't help with support for moveonly types. `toFuture` internally causes copies.

> Sure. I added this class because it proves to be challenging to implement a djinni interface that is based on Futures. Specifically, you have a bunch of intermediate results that are Futures, and you need to apply some processing on them when the caller invokes methods returning other Futures. I practice, I always have a bunch of Futures that internally depend on other Futures, in the form of a DAG. This means I needed additional resolved result storage solution for each Future. SharedFuture solved this issue, and I needed all these intermediate results to be SharedFutures, as well. I only ever produce regular Futures again when I finally export them out of the djinni boundary to another language.

To be honest, I can't really imagine it. Can you maybe produce a tiny toy example, so I can see "this works with SharedFuture and would be cumbersome without"? Sorry, I know I'm asking a lot, but I'm trying to understand.

Because what I'm imagining is something like this:
```c++
// These just do IO and suspend until ready
Future<FrameHeader> receiveFrameHeader();
Future<std::vector<std::byte>> receiveData(size_t size);

// This would be the interface method
// Frame image combines the image data with some metadata
Future<FrameImage> getNextFrameImage() {
    SharedFuture future{receiveFrameHeader()};

    // Count FPS
    future.then([](auto result) {
        result.get();
        countFps();
    });

    // Extract JPG
    return future
        .then([](auto result) {
            auto header = result.get();
            // are you trying to avoid this this slightly awkward pair that's needed to pass the header along?
            co_return std::make_pair(header, co_await receiveData(header.imageDataSize));
        })
        .then([](auto result) {
            auto [header, imageData] = result.get();
            co_return FrameImage{std::move(imageData), co_await receiveData(header.metadataSize)};
        };
}
```

PS: I'll likely be unable to respond for the next few days. If you and @LiFengSC agree that returning `SharedFuture` makes sense then feel free to just ignore my disagreements and merge it as-is.

Reply:
So the code is usually something like
```
SharedFuture<Something> inputA;
SharedFuture<SomethingElse> inputB;

void someInit() {
  SharedFuture<C> derived = something(inputA);
  _someMember = someLogic(inputA, derived);
  _otherMember = otherLogic(derived);
}

Future<Foo> someAccessor() {
  return someLogic(_someMember);
}

Future<Bar> otherAccessor() {
  return someLogic(_someMember, _otherMember);
}
```

So it's just cleaner if `then` produces SharedFuture by default, so I don't have to wrap each result into SharedFuture explicitly 90% of the time.

In any case, I updated the code to include both flavors.

Reply:
I don't have a strong opinion on this.  It kind of makes sense that once someone starts to use SharedFuture they'd want to keep using it.

Reply:
Thanks for updating the code. I've got no more complaints, looks great 🎉